src/qw/docsection.py

```diff
@@ -0,0 +1,708 @@
+    deeper levels, or a paragraph or list item together with all the
+    more indented list items directly below it.
+
+    One direction of iteration is with the `deeper` method. This
+    returns a new DocSection that iterates through the paragraphs in
+    this section but at a deeper outline level. If we are only
+    looking at one paragraph, `deeper` will return None as no deeper
+    iteration is possible.
+
+    The other direction of iteration is with the `next` method. This
+    moves the iteration to the next section in the document. If this
+    iterator was created with the `deeper` method `next` will not
+    iterate beyond the bounds of the section the original iteration
+    was pointing to; in this case, or if the iteration hits the end
+    of the document, `next` returns None.
+    """
+
+    def __init__(self, docx, from_index=None, outer_level=None, parent=None):
+        """
+        Create a DocSection as an empty object.
+
+        `next` should be called to initialize.
+        The section begins after the `after` element if set, or at
+        the start of the document if not.
+        The iteration ends at the end of the document or just before
+        the first paragraph that has a level equal to or lower than
+        outer_level (if not None)
+        """
+        self.docx = docx
+        bodies = docx.element.xpath("w:body")
+        if len(bodies) != 1:
+            msg = "not a single body"
+            raise QwError(msg)
+        # The "w:body" element of the document
+        self.element = bodies[0]
+        # The index of the first paragraph pointed to (within the
+        # body element)
+        self.start_index = from_index or 0
+        # The index just past the last paragraph pointed to
+        self.end_index = self.start_index
+        # The outline level (header level or list item indent)
+        # of the section pointed to.
+        self.start_level = None
+        # The outline level of the section that created this with
+        # its `deeper` method.
+        self.outer_level = outer_level
+        # The DocSection that created this DocSection with its
+        # `deeper` method.
+        self.parent = parent
+
+    def first_paragraph_text(self) -> str | None:
+        """Get the text in the first paragraph."""
+        if self.end_index == self.start_index:
+            return None
+        return self.element[self.start_index].text
+
+    def _get_number_level(self, element):
+        """
+        Get the number level.
+
+        This is found either in the paragraph formatting
+        or the number style, and refers to the indent level
+        of a list item.
+
+        We return the indent level plus 50; list items are
+        deeper than body text (at level 0).
+        """
+        bullets = element.xpath("w:pPr/w:numPr/w:ilvl")
+        if len(bullets) != 0:
+            val = bullets[0].attrib[ATTRIB_VAL]
+            if val.isnumeric():
+                return int(val) + 50
```

Comment:
could be worth setting 50 as a class field or constant?

Reply:
done